### PR TITLE
Add letter spacing for Android versions >= Lollipop (5.0).

### PR DESCRIPTION
**Pull request by zezhouliu**
Thanks for submitting a PR! Please read these instructions carefully:

- [x] Explain the **motivation** for making this change.
- [x] Provide a **test plan** demonstrating that the code is solid.
- [x] Match the **code formatting** of the rest of the codebase.
- [x] Target the `master` branch, NOT a "stable" branch.

## Motivation (required)
What existing problem does the pull request solve?

This PR makes enables `lineSpacing` on Android.  It is based on https://github.com/facebook/react-native/pull/9420, differing only with respect to minor library import references.

We have merged this into Airbnb's react-native fork: https://github.com/airbnb/react-native/pull/13

### Android fonts on Airbnb app
screen shot 2017-05-09 at 2 23 39 pm

### Android fonts specific use cases:

screen shot 2017-01-23 at 9 25 30 pm

screen shot 2017-01-23 at 9 27 42 pm

## Next Steps

Sign the [CLA][2], if you haven't already.

Small pull requests are much easier to review and more likely to get merged. Make sure the PR does only one thing, otherwise please split it.

Make sure all **tests pass** on both [Travis][3] and [Circle CI][4]. PRs that break tests are unlikely to be merged.

For more info, see the ["Pull Requests"][5] section of our "Contributing" guidelines.

[1]: https://medium.com/@martinkonicek/what-is-a-test-plan-8bfc840ec171#.y9lcuqqi9
[2]: https://code.facebook.com/cla
[3]: https://travis-ci.org/facebook/react-native
[4]: http://circleci.com/gh/facebook/react-native
[5]: https://github.com/facebook/react-native/blob/master/CONTRIBUTING.md#pull-requests

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java [hunk 1]
```diff
@@ -65,6 +68,21 @@ public void setText(ReactTextUpdate update) {
...
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+      float nextLetterSpacing = update.getLetterSpacing();
+      int fontSize = update.getFontSize();
+      if (!FloatUtil.floatsEqual(mLetterSpacing, nextLetterSpacing)) {
+        mLetterSpacing = nextLetterSpacing;
+        if(Float.isNaN(mLetterSpacing)) {
+          setLetterSpacing((float)0.0);
+        } else {
+          //calculate EM from proper font pixels
+          setLetterSpacing(1+(mLetterSpacing-PixelUtil.toDIPFromPixel(fontSize))/PixelUtil.toDIPFromPixel(fontSize));
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java]: Could you explain what this is doing?

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java [hunk 2]
```diff
@@ -65,6 +68,21 @@ public void setText(ReactTextUpdate update) {
...
       (int) Math.floor(update.getPaddingBottom()));
 
+    // API 21+: https://developer.android.com/reference/android/widget/TextView.html#setLetterSpacing(float)
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+      float nextLetterSpacing = update.getLetterSpacing();
+      int fontSize = update.getFontSize();
+      if (!FloatUtil.floatsEqual(mLetterSpacing, nextLetterSpacing)) {
+        mLetterSpacing = nextLetterSpacing;
+        if(Float.isNaN(mLetterSpacing)) {
+          setLetterSpacing((float)0.0);
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java]: style: space between the cast and 0.0 (like the setPadding call above)

**Comment by janicduplessis** [comment]
There's another similar PR for letter spacing, we might want to merge our efforts here to get it merged https://github.com/facebook/react-native/pull/13199

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java (hunk 2 above)
Comment (zezhouliu) [inline ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java]: thanks, I fixed this!

File: ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java (hunk 1 above)
Comment (zezhouliu) [inline ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java]: Added a comment inline, and pushed an update.

**Comment by zezhouliu** [comment]
Hi @ide: thanks for taking a look at this PR.

I made the changes that you suggested and also added an example to the explorer. I based it off https://github.com/facebook/react-native/blob/bd004568d3e006b61391417fc405e9483996dc81/RNTester/js/TextExample.ios.js#L387

I'm getting a CI error on travis: `'React/RCTBridge+Private.h' file not found`  Do you know how I can resolve this?

**Comment by sesm** [comment]
@zezhouliu @ide
Guys, what can community do to help you get this PR merged? 
What is the primary PR for this issue now, this one or #13199?

**Comment by RWOverdijk** [comment]
👍

**Comment by yenda** [comment]
@zezhouliu thanks for this PR ! Can you resolve the merge conflicts ?

@ide it seems that the changes you requested have been address is there still something ?

**Comment by shockdesign** [comment]
@zezhouliu Have you had a chance to update this? I was just working on my own version of this and realised you've already done the hard yards..

**Comment by motiz88** [comment]
Some issues with this PR as it stands (from the drive-by perspective of a RN user interested in this feature; I'm **not** a maintainer, or even a contributor yet - I can pretty much only semi-confidently _read_ the RN source code at the moment, but I have spent time [tinkering with letter spacing](https://github.com/motiz88/make-spaced-font) elsewhere):

- It appears to need a non-trivial update to be mergeable, in light of 6114f863c3aed826355530bcf404ee5aed2f927b in which a lot of the Text implementation has shifted.
- It seems to assume font sizes are `int`s (easily fixable in itself).
- It seems to hardcode an assumption that a `ReactTextUpdate` only has a single font size, which _may_ be at odds with Text nodes being nestable etc. Presumably a correct implementation would instead wrap the content in a custom Span that uses [`Paint.getTextSize()`](https://developer.android.com/reference/android/graphics/Paint.html#getTextSize()) to normalise a letter spacing value stored in points (DIP/SP depending on `allowFontScaling`) and pass it into
 [`Paint.setLetterSpacing(float)`](https://developer.android.com/reference/android/graphics/Paint.html#setLetterSpacing(float)).
- I am personally not sold on the points-to-em conversion formula (paraphrased for clarity):
  [`1 + (letterSpacingPoints - fontSizeDIP) / fontSizeDIP`](
https://github.com/zezhouliu/react-native/blob/ed7ce85fff154c64a861c72cdcadc9a5638d3f66/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextView.java#L82) - even putting aside DIP vs SP (`allowFontScaling` mentioned above), why doesn't it work out to 0 if `letterSpacing` is 0? But I could be missing something here.

**Comment by motiz88** [comment]
@zezhouliu Want to close this to keep the focus on #17398?

**Comment by hramos** [comment]
#17398 landed.

**Pull request closed without merging**